Add pinch zoom functionality

zbettenbuk opened the pull request.

zbettenbuk commented:
@jitsi/mobile please review and comment.

Please also note that I'm experimenting with a different approach avoiding passing the transform object in the props, but the movement limitation has a very difficult logic and makes it difficult to use Redux (nicely) for it. Anyway, stay tuned for a likely update.

zbettenbuk commented:
@paweldomas here is a fixed version of this prop-based solution. iOs now works fine.

I closed the other PR as we discussed that this should be the way to go, however I'll spend some time trying to test your suggestion about using only one component both for capturing the gesture and applying it to its children. If that works fine, I'll update this PR

zbettenbuk commented:
Jenkins, skip ci.

zbettenbuk commented:
@paweldomas looks final to me, feel free to review

paweldomas commented:
Hey @zbettenbuk 
So far I have only read the code without going into much details like styling etc. and tried the app on both Android and iOS. Here are few thing that I've noticed:
1. The transform is not preserved for each desktop stream when switching between two desktop streaming users. It seems weird given that the transformation is stored for stream id. Other weird thing is that the transformations are cleared on the conference left event and not when stream is removed. We need to answer a question here do we want to store it for each user/stream or do we want to preserve the transform only when the component unmounts ? If we store it for each stream then it should be saved/restored also when switching between the users and cleared when user leaves the conference (or stream is removed). If we don't want to store per user then it's fine to clear on conference left, but there's no need to store it for specific stream id.
2. The animation when switching between user videos is very slow. Was that something that changed in this PR or is it like that in general and I just haven't noticed it ?
3. There are glitches when switching between users. On Android the transform will kick in only after the animation is complete, like on this video [here].(https://drive.google.com/file/d/1L6kb03uJPPHGxCLMQGlK5NBxG8rLM3Z8qQ/view?usp=sharing)
On iOS I will first see the transformation kick into the old video that is fading out, and the screen shows after the animation is complete. It's kind of weird.
4. How are the drag limits supposed to work ? One time I see it allow to have only small stripe of blackness and the other time it allows to drag all the way up until only small stripe of screen is left visible on the screen. It seems to be related to phone orientation and size of the video. I'm trying to understand how it's supposed to work and if it's possible to eventually make it more consistent.

zbettenbuk commented:
@paweldomas thanks for the test and review, I'm addressing your comments - they are super valuable. The thing that the transform is not preserved between multiple desktop streams seems to be an indeed weird case, but it turned out that if you switch between 2 desktop streams the component is never unmounted - it only gets unmounted if you switch to an avatar view and back. So this needs some additional logic.

About the switching between videos: this seems to be related but it's quite difficult to reproduce. I'll do my best though.

Drag limit: The problem is that the stream is always full screen, even if only a smaller desktop stream is rendered. So the limitation is always constant (you can move 100points away from the edge of the screen), but when some parts of the screen is rendered black, it appears as if you could move more. I'll leave this problem to the very end of the PR as some changes may be needed to be introduced in the native RTCView component.

paweldomas commented:
>Drag limit: The problem is that the stream is always full screen, even if only a smaller desktop stream is rendered. So the limitation is always constant (you can move 100points away from the edge of the screen), but when some parts of the screen is rendered black, it appears as if you could move more. I'll leave this problem to the very end of the PR as some changes may be needed to be introduced in the native RTCView component.

Isn't it about the fact that the limit is in points and not % of the screen video's size ?

zbettenbuk commented:
> Yes, it is not unmounted. You will have to track 'stream' property updates and react there.

Yep, fixed!

paweldomas commented:
There's one other thing that has not addressed from my previous comment. Transform's lifetime is tied to the conference rather than to existence of a stream. It's being cleared on conference left rather than stream removed/participant left. I'm not strongly insisting on making it this way, but this could lead to problems in very large conferences or when there's a lot of different streams. It's very unlikely this will be a real problem, but I would like to know @lyubomir opinion, before approving it as is.

zbettenbuk commented:
> There's one other thing that has not addressed from my previous comment. Transform's lifetime is tied to the conference rather than to existence of a stream. It's being cleared on conference left rather than stream removed/participant left. I'm not strongly insisting on making it this way, but this could lead to problems in very large conferences or when there's a lot of different streams. It's very unlikely this will be a real problem, but I would like to know @lyubomir opinion, before approving it as is.

@paweldomas Right, I forgot to comment this: The only action I found was 'TRACK_REMOVED' and that is not really suitable to doing this. Is there a stream remove action too that contains the stream ID?

paweldomas commented:
>@paweldomas Right, I forgot to comment this: The only action I found was 'TRACK_REMOVED' and that is not really suitable to doing this. Is there a stream remove action too that contains the stream ID?

Yes, TRACK_REMOVED is what you need. Stream id comes from the video track if you check on the ParticipantView component.

Btw. I forgot to mention that, but I'm not sure if we should be putting transforms into the base/media feature. Maybe it should be in a separate one. I'm going to check with @lyubomir

zbettenbuk commented:
I take another look at TRACK_REMOVED and see if i can get some useful info feon that, thanks!

About the location of the feature: the transform component is physically located inside the Video conponent in the DOM tree, so i thought it belongs next to them, but i can be convinced otherwise

lyubomir commented:
@zbettenbuk,  does the commit '[RN] "The View is never the target of touch events"' make sense? I've extracted it from 'Add pinch zoom functionality' because (1) the modifications look to me to be self-sufficient and (2) they do make sense to me even if you're not thinking of "pinch to zoom." Could you please confirm?

lyubomir commented on react/features/base/media/middleware.js:
```diff
@@ -25,6 +30,9 @@ const logger = require('jitsi-meet-logger').getLogger(__filename);
     switch (action.type) {
+    case CONFERENCE_LEFT:
+        _clearVideoTransforms(store);
+        break;
```
1. Please separate `case` blocks with an empty line to visually aid the understanding of the fact that they are unrelated. I think I saw them bunched up in multiple files in this PR.

2. I presume it's for the sake of the abstraction's completeness but given all the other complexity around transforms I could've lived without the action `CLEAR_VIDEO_TRANSFORMS` I guess. The reducer could've directly reduced the actions which currently cause `CLEAR_VIDEO_TRANSFORMS` to be dispatched.

3. Why only `CONFERENCE_LEFT`? Why not `CONFERENCE_FAILED` as well?

4. Why clear all "video transform"s at the end of the conference? Why not clear when the stream associated with the "video transform" is no longer in the redux state? Or when the associated participant leaves?

zbettenbuk replied:
1. Ok, thanks!
2. Also, ok, I'll move them into the reducer, thanks!
3. Because I didn't know all the events and what they do and when. It would be good to have an extracted documentation about them. Maybe I'll spend some time having one.
4. Pawel also suggested this and I told him I take a look at that action, but as far as I checked (twice) there is no stream ID in that action. Am I missing something?

paweldomas replied:
>Pawel also suggested this and I told him I take a look at that action, but as far as I checked (twice) there is no stream ID in that action. Am I missing something?

There must be access to stream ID from TRACK_REMOVED event, have you checked good enough ? (I'll take a look now)

paweldomas replied:
There is a JitsiTrack in TRACK_REMOVED even which should be enough to get stream id...

zbettenbuk replied:
Found the stream.id in the jitsiTrack object, thanks! Btw it wasn't easy, as that's a huge object. Is there an easy way to find out what that object is consist of? E.g. some nice comments or an easy to read constructor?

zbettenbuk commented on react/features/base/media/components/native/VideoTransform.js:
```diff
@@ -0,0 +1,693 @@
+const maxOffset = 100;
+const maxScale = 5;
+const moveThresholdDismissesTouch = 2;
+const tapTimeoutMS = 400;
```
Fixed/added, thanks!

zbettenbuk commented on react/features/conference/components/Conference.native.js:
```diff
@@ -190,7 +190,6 @@ class Conference extends Component<Props> {
             <Container
                 accessibilityLabel = 'Conference'
                 accessible = { false }
-                onClick = { this._onClick }
```
1. Added explanations in VideoTransform#_onGesture and ParticipantView#render to explain these changes. Please review them, I'm happy to extend them if need be.
2. In order to extract that in a separate commit and not to loose functionality I'd need to implement a temporary solution to detect touches over the video (see previous explanations) without having the VideoTransform component in place (yet) - which I'd have to delete immediately when adding the VideoTransform component in the next commit. I'm happy to do this, but is it worth the effort?

lyubomir replied:
> In order to extract that in a separate commit and not to loose functionality I'd need to implement a temporary solution to detect touches over the video (see previous explanations) without having the VideoTransform component in place (yet) - which I'd have to delete immediately when adding the VideoTransform component in the next commit. I'm happy to do this, but is it worth the effort?

You've already developed a "clickable avatar" in this PR. Which doesn't really have to be an "avatar" and may instead be a higher-order "clickable". At the end of the day you've already identified two "views" which we may want to add "clickable" capabilities to so I'd say the chances are we'll find more. And with one stone we'd have hit a second bird: the unreadably formatted "clickable avatar" implemented in `ParticipantView`.

lyubomir replied:
Here's an initial `Pressable` React `Component` which I wrote in react/features/base/react/components/native/:

```javascript
// @flow

import React, { Component } from 'react';
import { TouchableWithoutFeedback } from 'react-native';

/**
 * The type of the React {@link Component} props of {@link Pressable}.
 */
type Props = {
    children: React$Node,

    /**
     * Called when the touch is released, but not if cancelled (e.g. by a scroll
     * that steals the responder lock).
     */
    onPress: Function
};

/**
 * Adds support for {@code onPress} to a child React {@link Component} (which
 * should probably not support the prop in question; otherwise, there's little
 * point of using {@code Pressable} then in the first place).
 */
export default class Pressable extends Component<Props> {
    /**
     * Implements React's {@link Component#render()}.
     *
     * @inheritdoc
     * @returns {React$Node}
     */
    render() {
        // onPress
        const { children, onPress } = this.props;

        if (onPress) {
            return (
                <TouchableWithoutFeedback onPress = { onPress }>
                    { children }
                </TouchableWithoutFeedback>
            );
        }

        // A Pressable without an onPress is a "no-op".
        return children;
    }
}
```

lyubomir replied:
@zbettenbuk, I've demonstrated the separation of the `onPress` modifications which prepare for "pinch-to-zoom" from the `VideoTransform`-related source code by implementing it. Let's please try to finish this PR by keeping the commits atomic (even if the next requires the previous).

lyubomir commented on react/features/base/media/components/native/Video.js:
```diff
@@ -68,28 +86,32 @@ export default class Video extends Component<*> {
+            const _zoomEnabled = zoomEnabled && type === 'desktop';
+
+            const objectFit = (_zoomEnabled && 'contain')
+                || 'cover';
```
That's `const objectFit = _zoomEnabled ? 'contain' : 'cover';`

lyubomir commented on react/features/base/media/components/native/Video.js:
```diff
@@ -68,28 +86,32 @@ export default class Video extends Component<*> {
-            const style = styles.video;
-            const objectFit = (style && style.objectFit) || 'cover';
+            // Actual zoom is only enabled if the stream
+            // is also a desktop stream.
```
Please utilize the full maximum of 80 characters for comments.

lyubomir commented on react/features/base/media/reducer.js:
```diff
@@ -87,6 +90,11 @@ const VIDEO_INITIAL_MEDIA_STATE = {
+        return {
+            ...state,
+            transforms: {}
+        };
```
I guess there's another `case` which needs to be separated with a new line from the totally unrelated `case` bellow it.

zbettenbuk replied:
I'm not sure what you mean here. Can you please explain?

lyubomir replied:
It's another comment that I fixed myself eventually.

lyubomir commented on react/features/base/participants/components/ParticipantView.native.js:
```diff
@@ -218,20 +234,34 @@ class ParticipantView extends Component<Props> {
+                            }
+                        </View>
+                    </TouchableWithoutFeedback>)
+                    || this._renderAvatar()) }
```
That expression is plain unreadable with its current formatting.

lyubomir commented on react/features/base/media/reducer.js:
```diff
@@ -138,3 +149,27 @@ ReducerRegistry.register('features/base/media', combineReducers({
+ * @returns {Object}
+ */
+function _storeVideoTransform(state, action) {
+    const { streamId, transform } = action;
```
That's `function _storeVideoTransform(state, { streamId, transform }) {`.

lyubomir commented on react/features/base/media/reducer.js:
```diff
@@ -138,3 +149,27 @@ ReducerRegistry.register('features/base/media', combineReducers({
+        transforms = {};
+    }
+
+    transforms[streamId] = transform;
```
Well, that's technically a modification of the redux state. Though you do get state and compute a next state in order to not modify `state`'s props, you're still modifying the very `transforms` object referenced by `state`.

lyubomir commented on react/features/base/media/components/native/VideoTransform.js:
```diff
@@ -0,0 +1,732 @@
+/**
+ * The default/initial transform (= no transform).
+ */
+const defaultTransform = {
```
How is `defaultTransform` different from `MIN_SCALE`, etc. which you converted from `minScale`? Is it not being used as a literal?

zbettenbuk replied:
The MIN_SCALE just happens to be equal to defaultTransform.scale for now, but they are not related. We can reduce MIN_SCALE to below 1 (if we want to allow the user to be able to scale under 1x magnification) but the defaultTransform.scale will still remain 1, as that means no-op.

lyubomir replied:
I'm saying that 'defaultTransform' appears to be a "literal" just line "MIN_SCALE" so I don't understand why you applied the appropriate naming to "MIN_SCALE" but didn't change `defaultTransform` to `DEFAULT_TRANSFORM`. Am I missing something?

zbettenbuk replied:
Ah, sorry, I misunderstood then. Yeah, I fix that, thanks!

lyubomir commented on react/features/base/media/components/native/styles.js:
```diff
@@ -6,6 +6,14 @@ import { ColorPalette } from '../../../styles';
+    /**
+     * Base style of the transformed video view.
+     */
+    transformedView: {
```
I'm really confused by the naming around "video transforms". The action types and creators, the class `VideoTransform` talk about an abstraction "video transforms". But I see these `transforms` in the redux state `base/media` that are merely "transforms". And I was thinking that actually "transforms" are like a base/super and then there are "video transforms". But then there's no such mechanism. Such a confusion is preemptively addressed by our coding style: one name per abstraction. In accord with it I'd rather go with `videoTransforms` in the redux state, `videoTranformedView` (maybe).

zbettenbuk replied:
I renamed this style class, thanks.

Not sure however what you mean here exactly. I used the word "transform(s)" everywhere in the VideoTransform since the name of the component predefines that it's all about video. On the other hand I couldn't imagine a situation where we (visually) transform anything other than a video stream. Or did I maybe misunderstand your comment?

lyubomir replied:
There's "transforms" in the redux state of "base/media". But "base/media" contains both audio and video-related state. So "transforms" is unclear to me.

zbettenbuk replied:
The transforms object is in the video subtree of the state of the feature, as it's being reduced by the video reducer that is later combined with the audio reducer to handle the base/media tree.

lyubomir commented:
@zbettenbuk,  I added a third commit "Add pinch zoom functionality (Coding style, prevent an unintentional …" to demonstrate some of the changes that I just described as comments.

lyubomir commented on react/features/base/participants/components/ParticipantView.native.js:
```diff
@@ -218,20 +234,34 @@ class ParticipantView extends Component<Props> {
-                        size = { this.props.avatarSize }
-                        uri = { avatar } /> }
+                    && ((onPress && <TouchableWithoutFeedback
+                        onPress = { onPress }>
```
There's another problem with the `onPress`:
1. When the `TintView` is rendered, pressing stops working/toggling the toolbox.
2. Pressing on the "connectivity info label" doesn't toggle the toolbox.

zbettenbuk replied:
Not sure now if you fixed this with the recent commit or it's still an issue. I tested the connectivity issue and pressing the screen when the spinner is on seems to work fine (i.e. brings up the controls).

Which one is the connectivity info label?

lyubomir replied:
I have solved the regressions. The connectivity info label is visible in the source code.

lyubomir commented on react/features/large-video/components/LargeVideo.native.js:
```diff
@@ -107,16 +112,19 @@ class LargeVideo extends Component<Props, State> {
                 onDimensionsChanged = { this._onDimensionsChanged } >
                 <ParticipantView
                     avatarSize = { avatarSize }
+                    onPress = { onPress }
```
In this file `onPress` is decomposed out of `this.props` and use on this line only. On the next line we have `this.props._participantId`. Let's have `this.props.onPress` then.

zbettenbuk replied:
Right, I didn't see that sorry. However if we decompose nearly everything above 1 variable (see this.state a couple of lines above) then I guess it's more consistent to decompose props too so then I'd rather change _participantId than onPress. What do you think?

lyubomir replied:
Whatever, just be consistent.

zbettenbuk commented on react/features/base/participants/components/styles.js:
```diff
@@ -12,6 +12,19 @@ export default createStyleSheet({
+        flex: 1,
+        justifyContent: 'center'
+    },
+
```
How do you mean useless?

lyubomir replied:
Please grep the source and see that as it stands today neither of these two styles are used. As I said, they need to be removed.

zbettenbuk commented on react/features/base/participants/components/styles.js:
```diff
@@ -40,6 +53,7 @@ export default createStyleSheet({
         flex: 1,
-        justifyContent: 'center'
+        justifyContent: 'center',
+        overflow: 'hidden'
```
To avoid rendering off screen, e.g. in PiP

lyubomir replied:
Then it's totally unrelated to "pinch to zoom" and needs to go into a separate commit. A description in the commit will also be nice because it's not really obvious to everyone why that's done.

zbettenbuk replied:
No, it's related, because it's the transformed view that gets rendered off-screen when other than default transform is applied

zbettenbuk commented on react/features/base/media/components/native/Video.js:
```diff
@@ -77,29 +86,29 @@ export default class Video extends Component<*> {
-            const objectFit = (style && style.objectFit) || 'cover';
+            // Actual zoom is only enabled if the stream is also a desktop
+            // stream.
+            const _zoomEnabled = zoomEnabled && type === 'desktop';
```
Are you sure this is a right logic? What if someone wants a simple video view with no controls popping up on it (ie no onPress) but still wants pinch&zoom? I think connecting the availability of p&z to the presence of onPress is not a straightforward idea

zbettenbuk replied:
Moved the desktop logic to (Abstract)VideoTrack so then the Video can remain as generic as possible. If we ever want to be able to zoom on non-desktop stream, we need to smarten (or remove) this logic. What do you think?
